Update and extend governance [WIP] (#4980)
pymc-devs/pymc: GOVERNANCE.md

# Main Governance Document

## The Project

The PyMC Project (The Project) is an open source software project

> affiliated with the 501c3 NumFocus Foundation. The goal of The Project is to

Reviewer: Nit on the stylization of NumFOCUS

```suggestion
affiliated with the 501c3 NumFOCUS Foundation. The goal of The Project is to
```

develop open source software and deploy open and public websites and services
for reproducible, exploratory and interactive computing. The Software developed
by The Project is released under OSI approved open source licenses,
developed openly and hosted in public GitHub repositories under the
[GitHub organization](https://github.com/pymc-devs). Examples of
Project Software include the PyMC3 library and its documentation, etc.
The Services run by the Project consist of public websites and web-services
that are hosted at [http://docs.pymc.io](http://docs.pymc.io)

The Project is developed by a team of distributed developers, called
Contributors. Contributors are individuals who have contributed code,
documentation, designs or other work to one or more Project repositories,
or who have done significant work to empower the Community,

> moderating on [Discourse](http://docs.pymc.io),

Reviewer 2: This link does not point to the discourse.

Also what's our correct docs link these days?
docs.pymc.io and pymc3.readthedocs.io don't show the same things.
(We should set up forwarding I guess?)

Author: > This link does not point to the discourse.

Thanks, will fix that

> Also what's our correct docs link these days?
docs.pymc.io and pymc3.readthedocs.io don't show the same things.
(We should set up forwarding I guess?)

Docs link is and should be docs.pymc.io. We'll make rtd serve the docs on that same url with the v4 release, and by "we" I mean Sayam who is the rtd wizard

organizing [PyMCon](https://pymcon.com) or helped on other platforms and events.
Anyone can be a Contributor. Contributors can be affiliated with any legal

> entity or none. Contributors participate in the project by submitting,

Reviewer: > Contributors participate in the project by submitting, reviewing and discussing GitHub Pull Requests and Issues and participating in open and public Project discussions on GitHub, Slack, Gitter chat rooms and mailing lists.

This sentence seems contradictory to the idea of envisioning contributions as more than just code and code-related things - from previous sentence, organizing PyMCon and moderating Discourse are valid and valuable contributions but don't have anything to do with code. Perhaps we should delete this?

Reviewer 3: > It needs to be easier for people to become recurring contributors - while joining the Slack (and thus becoming a recurring contributor, under this document's definition) is a huge milestone in a person's involvement with PyMC, it's a high bar to expect first-time contributors to meet to graduate into recurring-contributor-hood.

big +1 here

Reviewer 3: also we can start to invite people to a few of the slack channels and later if they stay active we add them to more channels. but yes i agree so much with this entire comment

# Main Governance Document

## The Project

The PyMC Project (The Project) is an open source software project
affiliated with the 501c3 NumFOCUS Foundation. The goal of The Project is to
develop open source software and deploy open and public websites and services

> for reproducible, exploratory and interactive computing. The Software developed

Reviewer 4: Should probably say something about scientific/statistical computing, since that is the major part of the goal.

For the nomination to go forward, it has to be ratified by the Steering Council.
For a nomination to be rejected, clear reasoning behind the decision must be
shared with the rest of the team. People whose nomination has been rejected can
be nominated at any time again in the future, three months after the previous
nomination at the earliest. The nomination process is explained below
in more detail in a section of its own.

Interns and contractors are added to the team as Recurrent Contributors.
We consider the selection/hiring process to replace the nomination process.

#### Current Recurring Contributors
Contributors who are also part of a dedicated team or are institutional
contributors will have so indicated after their name.
Dedicated teams only cover a small part of the work needed to
get the project going, tasks like fundraising, outreach and marketing,
or organizing events for example don't (yet) have a dedicated team.
Contributors don't need to be part of any dedicated team.

- Abhipsha Das (docs)
- Benjamin Vincent (docs - PyMC Labs)
- Bill Engels (dev)
- Jon Sedar
- Kaustubh Chaudhari (dev)
- Larry Dong (dev)
- Lorenzo Toniazzi (docs)
- Olga Khan (docs)
- Peadar Coyle

> - Raul Maldonado (docs)

Reviewer 4: Do we want to be maintaining this list here? You can always identify members by looking at Github activity, no?

Author: > Do we want to be maintaining this list here?

I really want to maintain this list. The **only** con I see to the list is maintenance burden which in my opinion is clearly outweighted by all pros. 

Listing people here makes new team members feel part of the team, so far I have yet to find someone who is new to the team but has reservations about this section. It also clearly signals that we value our team members and want them to be recognized (even if they mostly work on fundraising and public relations on twitter and medium). It will log the history of members which at the same time will serve as an indicator to prospective team members that we are an active team continuously onboarding people

> You can always identify members by looking at Github activity, no?

Definitely not, and saying so would be going against ourselves as we have said above in multiple occasions that any and all contributions are welcome, not only code or pull requests. The most similar proxy are the members in the pymc organization, but that doesn't always show everyone.

I really really think that this will be good for the project and won't be hard to maintain. And if it turns out I'm wrong we can just come here a year from now, delete it and be done with it.

- everyone on slack! This may be too long so it could be a good idea to
format as a table instead or to put within a `<details>` tag so

> it's hidden by default while still available publicly

Reviewer 2: Here's a ToDo before merging.

Author: yes, I am also a bit unclear about how to better address that, should I again tag everyone with a "make sure you are listed here if you want to continue being part of slack"? List everyone and be done with it? Do nothing for now and have them follow the recurrent contributor nomination process as if being in slack was an automatic nomination but not an automatic membership card? 

There are many people on slack with whom I have never interacted nor I think they have been active at all in the last 1-2 years.

Reviewer 2: My opinion about this is to not keep a public list for recurring contributors.

Author: I did not express myself very well before, but the way I see it, the list of recurring contributors being public or even the existence of such list altogether is a rather insignificant part of everything that hides behind the questions above.

From what I understand, this governance defines being accepted as recurrent contributor as the requirement to join slack. 

If that is the case, it leaves us with the questions above, should everyone on slack be a recurrent contributor? Why out of the gsoc participants who did not contribute after gsoc finished (to the best of my knowledge) only some are still on slack? Did they leave voluntarily (I for once have no idea how to leave slack workspaces, only how to log out) or were they removed due to lack of activity?

If that is not the case, then we need to define how/when/why people are added or removed from slack. Because I do not know the answer and the "not being rules about it" basically ends up meaning that in practice the answer is that to join slack (or the project) one needs to be friends with the core contributors or well connected. And this will not do.

## Core Contributors
Core Contributors are those individuals entrusted with the development and
well being of the Project due to their frequency of quality contributions over
a sustained period of time. They are the main governing and decision body
of the Project and are therefore given voting and managing rights to the Project
services (i.e. commit rights on GitHub or moderation rights on Discourse).
The exact permissions of all Core Contributors may not be the same
and depend on their team memberships. Even if they have commit rights,
Core Contributors should still have their pull requests reviewed by at least
one other Core Contributor before merging unless prevented by a major force

> reason. If overstepping, Core Contributors can also be subject to a vote

Reviewer 5: I'm not sure what "a major force reason" means

Author: "prevented" might not be the right word for what I was trying to say. The main example that comes to mind is imagine we somehow break our docs, or an external dependency does and I see the fix very clearly, send a PR and it passes CI and everything. I'd start with a `@here` on slack to try and get reviews fast and merge but if after 2-3 hours it were not merged yet then I'd probably merge myself without any other aproval or review. Even more extreme similar scenario, if the issue happened during a pydata sprint I was hosting by myself I'd do my best to fix it in a matter of minutes with or without extra reviews. I hope it makes sense, any wording improvements will be extremely welcome

Reviewer 4: maybe something along the lines of: "... before merging. In rare circumstances, a Core Contributor may bypass this review if in their best judgement it is appropriate to do so, but such expedited PR merges must be justifiable and ultimately subject to review post hoc, if necessary."

Reviewer 5: Yeah, this makes sense, thank you! I would even aim weaker here, something like 
>Core Contributors should still have their pull requests reviewed by at least one other Core Contributor, though they may use common sense in case of extenuating circumstances.

Author: Thanks both for making this more clear! I think it will be the first time I will use extenuating in English! :smile:

Core Contributor. For the nomination to go forward, it has to be
ratified by the Steering Council.
For a nomination to be rejected, clear reasoning behind the decision must be
shared with the rest of the team. People whose nomination has been rejected can
be nominated at any time again in the future, three months after the previous
nomination at the earliest.

### Current Core Contributors
Contributors who are also part of a dedicated team or are institutional
contributors will have so indicated after their name.
Dedicated teams only cover a small part of the work needed to
get the project going, tasks like fundraising, outreach and marketing,
or organizing events for example don't (yet) have a dedicated team.
Contributors don't need to be part of any dedicated team.

* Adrian Seyboldt (dev - PyMC Labs)
* Alex Andorra (dev - PyMC Labs)
* Austin Rochford
* Brandon T. Willard (dev)
* Chris Fonnesbeck (dev, docs)
* Colin Carroll (dev)
* Eelke Spaak (dev)
* Eric Ma (dev - PyMC Labs)
* George Ho (dev)
* Junpeng Lao (dev)
* Luciano Paz (dev - PyMC Labs)
* Marco E. Gorelli (dev, docs)
* Martina Cantaro (docs)
* Maxim Kochurov (dev - PyMC Labs)
* Meenal Jhajharia (dev, docs)

> * Michael Osthege (dev, docs)

Reviewer 2: I know too little about docs to be listed that way
```suggestion
* Michael Osthege (dev)
```

While this approach has served us well, as the Project grows and faces more
legal and financial decisions and interacts with other institutions, we see a
need for a more formal governance and organization model.
We view this governance model as the formalization of what we are already doing,
rather than a change in direction.

## Community and Team Architecture
The PyMC community is organized in an onion-like fashion.
The tiers relevant to the project governance are listed below sorted by
increasing responsibility. Due to the onion-like structure, members of a group are
also members of all the groups listed above:

* Contributors
* Recurring Contributors
* Core Contributors

> * Steering Council

Reviewer 2: Does the steering council have a min/fixed/max size?

Author: I'd set some constraints, but I wasn't really sure which ones. Sorry I'm a bit all over the place, this is from the PR description:

> council membership constraints
>
> I think we should have a minimum number of members, say 5, with at least one member per team and at most 2 institutional contributors of the same company. Or some other numbers but some kind of constraints on those ends. I don't care much about the numbers really, but I think some constraints would be good to have

Author: Trying to be more specific, one proposal could be:

* between 4-7 members
* at least 1 member per team
* at most 2 institutional members per company

* BDFL

Recurring Contributors comprise what we understand as the PyMC Team.
The Team will generally act as a single unit, except for some specific
questions where dedicated teams will prevail.

> The PyMC project currently counts with the Developer and Documentation teams.

Reviewer 6: ```suggestion
Currently, there are two teams within the PyMC project, the Developer and Documentation teams.
```

Team members can be part of one, some or none of these dedicated teams.

> ![community_diagram](docs/community_diagram.png)

Reviewer 4: Question: can you be a core contributor without being part of either the dev team or the doc team? If not, the diagram would need some tweaking.

Author: Yes, one can be core contributor or steering council member without being on any of the teams

they will be considered for removal from the Council. Before removal, inactive
Member will be approached by the BDFL to see if they plan on returning to
active participation. If not they will be removed immediately upon a Council
vote. If they plan on returning to active participation soon, they will be
given a grace period of six months. If they don’t return to active participation
within that time period they will be removed by vote of the Council without
further grace period. All former Council members can be considered for
membership again at any time in the future, like any other Core Contributor.
 Retired Council members will be listed on the project website, acknowledging
the period during which they were active in the Council.

The Council reserves the right to eject current Members, other than the BDFL,
if they are deemed to be actively harmful to the project’s well-being, and
attempts at communication and conflict resolution have failed.

<!-- todos:

- vote of no conficence. Either core or recurring contributors
should also be able to remove people from the council, even if by a very high
majority vote. IMO (Oriol) this point is not up for discussion in itself, only in

> how it should happen.

Reviewer 2: * How is such a vote initiated?
* What is the voting mechanism?
  * Attendance meeting vs. asynchronous
  * Minimum number of non-neutral votes (or similar)
  * simple/50/66/75... % majority required?

Author: We can copy that from ArviZ: https://github.com/arviz-devs/arviz/blob/main/GOVERNANCE.md#vote-of-no-confidence

Reviewer 6: The ArviZ section on votes of no confidence sounds good to me, with one exception: if a vote of no confidence for a single Council Member is called by Council Members, who then decide to ask the whole Core Community to vote, it seems to me still a two thirds majority would be needed. As it's phrased now, in that case, only a simple (half) majority is needed.

(Example: let's say 3/10 Council Members want one Council Member removed (but 7/10 do not), if those three Council Members then call the full Core Team to vote, can they bypass the needed two thirds majority typically required for votes of no confidence? Seems like a loophole we do not want. Very unlikely of course, but easy to fix.)

- election discussion. Do we want to keep the current council votes new council

>   or should core/recurring contributors vote?

Reviewer 2: IMO there should be a size limit to the steering council.
If it's a small council, I think the core contributors should vote; otherwise with a big steering council it probably wouldn't make a big difference in practice?

Reviewer 6: Agreed, smaller size is more efficient. And most decisions are made with the full core community anyway (order of dozens of people).

I'd be in favour of core community voting steering council, rather than steering council votes steering council (but not "simple" recurring contributors). But still current steering council can propose new council and express preference etc., of course.

> - Should there be a minimum of 1 council member from each dedicated team?

Reviewer 2: 👍

- Make decisions about specific technical issues, features, bugs and pull
  requests.
- Make decisions about the Services that are run by The Project and manage
  those Services for the benefit of the Project and Community.
- Make decisions when regular community discussion doesn’t produce consensus
  on an issue in a reasonable time frame.

### Current Steering Council

The current Steering Council membership comprises:

- void until we merge this and then have an election.
  The process for renewing and electing a new member is the same so it won't
  make any difference.

Institutional Council Members are indicated as `name (company)`

### Council membership

To become eligible for being a Steering Council Member an individual must be a
Core Contributor who has produced contributions that are substantial in
quality and quantity, and sustained over at least one year. Potential Council
after asking if the potential Member is interested and willing
to serve in that capacity.

Similarly to when nominating new team members, when considering potential
Council Members one should look at candidates with a
comprehensive view of their contributions. This will include but is not limited

> to code, code review, infrastructure work, mailing list and chat participation,

Reviewer 2: ```suggestion
to code, code review, infrastructure work, forum and chat participation,
```

The Steering Council and its Members play a special role in certain situations.
In particular, the Council may:

- Make decisions about the overall scope, vision and direction of the
  project.
- Make decisions about strategic collaborations with other organizations or
  individuals.
- Make decisions about specific technical issues, features, bugs and pull
  requests.
- Make decisions about the Services that are run by The Project and manage
  those Services for the benefit of the Project and Community.
- Make decisions when regular community discussion doesn’t produce consensus
  on an issue in a reasonable time frame.

> 

Reviewer 4: May want to add something like: "Make decisions about strategic financial issues." 

So, not day-to-day decisions like approving reimbursements, but those concerning the use of money in a large-scale sense.

Author: I was acutally considering reducing the list or downplaying it a little. While I expect council members to hold a lot of power and probably effectively decide the roadmap and strategy of the project. From the paragraph above (which I barely modified) I think they should do that as core contributors, which should be the main governance body, with the steering council intervening only if necessary. After all, council members are core contributors with at least a year of significant contributions and well respected enough by the rest of core contributors to be elected as council members.

### Current Steering Council

The current Steering Council membership comprises:

> - void until we merge this and then have an election.

Reviewer 5: this looks great to me (and i'm happy to "resign" if it helps), but I think the members to the left _technically_ remain the steering council until the election, if i'm reading the old/existing governance document right.

Author: yes, that is the comment after that sentence. As we are setting a maximum of 2 years and the last election was longer ago than that and there are no differences between being elected or reelected, the upcoming election will effectively be as if the steering council was void.

Reviewer 5: thanks for explaining!

  The process for renewing and electing a new member is the same so it won't
  make any difference.

Institutional Council Members are indicated as `name (company)`

### Council membership

To become eligible for being a Steering Council Member an individual must be a
Core Contributor who has produced contributions that are substantial in
quality and quantity, and sustained over at least one year. Potential Council
after asking if the potential Member is interested and willing

> to serve in that capacity.

Reviewer 4: Check the grammar here.

specifically called for. Private subcommittee communications should happen on
the main private mailing list of the Council unless specifically called for.

> Even if the BDFL does not sit on a specific subcommittee, they still retains

Reviewer 2: ```suggestion
Even if the BDFL does not sit on a specific subcommittee, they still retain
```

specific aspects of the project. Like the Council as a whole, subcommittees
should conduct their business in an open and public manner unless privacy is
specifically called for.

Even if the BDFL does not sit on a specific subcommittee, they still retain
override authority on the subcommittee's decisions. However, it is expected that
they will appoint a delegate to oversee the subcommittee's decisions, and
explicit intervention from the BDFL will only be sought if the committee
disagrees with the delegate's decision and no resolution is possible within the
subcommittee. This is a different situation from a BDFL delegate for a specific
decision, or a recusal situation, in which the BDFL gives up their authority
to someone else in full.

### NumFOCUS Subcommittee

The Council will maintain one narrowly focused subcommittee to manage its
interactions with NumFOCUS.

- The NumFOCUS Subcommittee is comprised of 5 persons who manage project
  funding that comes through NumFOCUS. It is expected that these funds will
  be spent in a manner that is consistent with the non-profit mission of
  NumFOCUS and the direction of the Project as determined by the full

>   Council.

Reviewer 4: Yeah this pretty much addresses what I suggested above. I like it.

Author: cool, I think I did not modify that part and git is messing up actually :smile:

- This Subcommittee shall NOT make decisions about the direction, scope or

>   technical direction of the Project.

Reviewer 4: It would not even make decisions regarding the financial direction of the project. The spending decisions should be consistent with directions set out by the steering council.

specifically called for. Private subcommittee communications should happen on
the main private mailing list of the Council unless specifically called for.

Even if the BDFL does not sit on a specific subcommittee, they still retain
override authority on the subcommittee's decisions. However, it is expected that
they will appoint a delegate to oversee the subcommittee's decisions, and
explicit intervention from the BDFL will only be sought if the committee
disagrees with the delegate's decision and no resolution is possible within the
subcommittee. This is a different situation from a BDFL delegate for a specific
decision, or a recusal situation, in which the BDFL gives up their authority
to someone else in full.

> ### NumFOCUS Subcommittee

Author: do we need this subcommittee? cc @fonnesbeck I don't really know exactly what it's supposed to do from the description, (it looks like an admin body mostly to execute what the team decided?), if it was active when created or if you expect it to be active now with CZI for example, so I have not touched anything for now

Reviewer 4: NumFOCUS requires this group. We could make it more useful by assigning them the responsibility of approving specific expenditures, like travel, conferences, and other uses of PyMC money, which ultimately flows through NumFOCUS anyhow.

- Jon Sedar
- Thomas Wiecki

## Core Contributors
Core Contributors are those individuals with commit or managing rights to any of
the repositories or services of the PyMC Project. However, the exact permissions
of all Core Contributors may not be the same and depend on their team memberships.
Therefore, Core Contributors may act as Recurrent Contributors depending

> on the task at hand.

Reviewer 2: That sounds like the "core contributors" are not a GitHub Team?
> Core Contributors are those individuals with commit or managing rights to **any** of
the repositories

Phrased like this, any PyMC3 or Docs Team member is automatically a core contributor.

> Therefore, Core Contributors may act as Recurrent Contributors depending
on the task at hand.

I'm not sure what this means

Author: > That sounds like the "core contributors" are not a GitHub Team?

No, I have tried to make everything less code centered and less GitHub centered. In my opinion, we could potentially have core contributors or council members who don't even have a GitHub account if they were for example Discourse admin or moderator.

> Phrased like this, any PyMC3 or Docs Team member is automatically a core contributor.

We probably need to fix the wording. What I was trying to phrase is that someone with write permissions to pymc-examples is a core contributor, even if they don't have write permissions on pymc3 repo, or like I said above, no write permissions on github but moderator on discourse. Members of dev or doc team don't necessarily have write permissions to any repo, teams are comprised of core and recurrent contributors, only core contributors would have write permission.

> I'm not sure what this means

This links to the previous point, and is mostly being sensible but I included that in hopes of being transparent and explicit. An example of that would be that while someone might be a core contributor with write permissions to pymc3 (part of dev team), they should not merge a doc only PR without some approval from a doc team member (depending on the extent and complexity of the PR)

Reviewer 6: > Core Contributors are those individuals with commit or managing rights

If you want to explicitly disentangle Core Contributor membership from specific services (code/docs, etc.) then I would not start with this.

We believe this to be necessary given the current size and scope of
the PyMC project. Contributors may be proficient in some areas of the
project while being inexperienced in some others.

### Core Contributor membership
To become a Core Contributor, one must already be a Recurring Contributor.
Core Contributors can nominate any Recurring Contributor to become a
Core Contributor. For the nomination to go forward, it has to be

> ratified by the Steering Council and the BDFL.

Reviewer 2: The BDFL chapter above doesn't read like the BDFL needs to be involved in such decisions.
> Because rarely used, we refer to BDFL’s final authority as a “special” or “overriding”
vote.

Veto yes, but the "ratified by [...] the BDFL" reads like its a must.

Author: We can remove BDFL from here, I agree 100% with your point and only wanted that to mean "ask to ensure there is no veto"

For a nomination to be rejected, clear reasoning behind the decision must be
shared with the rest of the team. People whose nomination has been rejected can
be nominated at any time again in the future, three months after the previous
nomination at the earliest.

### Current Core Contributors

<!-- we could list them here or link to the pymc-devs organization, however,
linking to the pymc-devs org would require the core team not being secret.
I actually don't know if this is already the case, if it's possible even nor
obviously how it would be done. I do believe it's important to be clear and
explicit about who are core contributors, especially for people interested
in joining the project to reach out, but also for organizations and meet-ups

> to ask for speakers.

Reviewer 2: Just checked it: The **maximum visibility setting for a team is to be visible within the organization**.

Outsiders can still learn about a team's existence if it's mentioned, but **only org members** can mention or inspect a "visible" team.

Author: Thanks! That means that it's either listing here or the team of core contributors not being public

Regarding the reaching out, we do have the project mail and some people
may not desire being contacted privately. I also think we should have some
rules or guidance on this here and/or in the community/contributing pages.-->

## Recurring Contributors
Recurring Contributors are those individuals who can participate from
The Project private communication channels.

### Recurring Contributor membership
Recurring Contributors can nominate any Contributor to participate in the
Project private communication channels and become a Recurring Contributor.
For the nomination to go forward, it has to be ratified by

> the Steering Council and the BDFL.

Reviewer 2: same as above

For a nomination to be rejected, clear reasoning behind the decision must be
shared with the rest of the team. People whose nomination has been rejected can
be nominated at any time again in the future, three months after the previous
nomination at the earliest.

#### Current Recurring Contributors

- everyone on slack! This may be too long so it could be a good idea to
format as a table instead or to put within a `<details>` tag so

> it's hidden by default while still available publicly

Reviewer 2: Three levels of membership (council/core/recurring) are a lot to manage already. I don't think we need to keep a list of recurring contributors, since there are no formal rights/obligations with that status. (Other than being on the Slack channel?)

Author: While recurrent contributors don't have many rights nor obligations by design<sup>1</sup> one of the goals in "creating" the position is making new contributors feel and be part of the team earlier on with the hopes of attracting and retaining more people, and one general goal of the governance update was to be more clear and transparent.

I'd suggest we try having this here and if at some point it gets outdated or is too much work for us.

1: they only participate from private channels, slack and lab meetings now, and can nominate people for recurrent contributors, but they could also vote steering council members even if they were not eligible :thinking:

Reviewer 4: I agree that tracking recurring contributors is not something we should invest much (any) time in.

Author: I think especially for recurrent contributors it will be important to a lot of people and will help new members feel part of the team more easily. I suggest we try it (we can make new members add themselves here so we only have to merge a PR) and see if it becomes too much work.

Reviewer 3: i’d say it’s worth trying because i don’t know with all the new people joining slack in docs team etc, we can make them feel like they’re a real part of the team and not guests this way. i think we can find a middle ground. the goal remains: make it easier for people to contribute + accessible/welcoming to join the community. so i would say we do the safe and easier thing of: once people contribute and we think they'll be helpful in discussions we invite some of them to slack -> if we see that they contribute for n number of weeks more (this can be informal, say 3-4 weeks or 2-3 important PRs) then we give them triage permissions. so they're a part of the organization, but because they're new people they don't get write permissions. i also agree with oriol's idea "I suggest we try it (we can make new members add themselves here so we only have to merge a PR) and see if it becomes too much work." this sounds good as well. either way, the exact method doesnt matter, we dont need to formally track them, but just things we can do to make people feel like an active part of the community. like it was after chris invited me to slack that i got invested in discussions and meetings and i think we would want that, people who are genuinely invested in pymc

## Conflict of interest

It is expected that the BDFL, Council Members and Contributors will be
employed at a wide range of companies, universities and non-profit organizations.
Because of this, it is possible that Members will have conflict of interests.
Such conflict of interests include, but are not limited to:

- Financial interests, such as investments, employment or contracting work,
  outside of The Project that may influence their work on The Project.
- Access to proprietary information of their employer that could potentially
  leak into their work with the Project.

All members of the Council, BDFL included, shall disclose to the rest of the
Council any conflict of interest they may have. Members with a conflict of
interest in a particular issue may participate in Council discussions on that
issue, but must recuse themselves from voting on the issue. If the BDFL has
recused themselves for a particular decision, they will appoint a substitute
BDFL for that decision.

## Team Organization
<!-- Open questions

1. do we want to give commit rights to core contributors to some repositories

> only based on the team they are part of?

Reviewer 2: Yes, IMO there should be no "core-devs" GitHub team, but just {pymc3,examples}-{Triage,Write,Admin}.
Possibly even leaving the Triage level out and replacing it with proper branch protection (review required before merging), because Triage doesn't give permission to cancel/re-run CI jobs, but that's something we often need..

Plus a steering council Team that has Admin level on everything

Author: That sounds like a sensible mapping of our structure to github. I wanted to keep everything github agnostic to begin with, once the tiers look decided we can add that to the doc

(quick discussion with myself - Oriol) With pymc3 being the source of the
website and a large part of the docs, both members of developer and documentation
team need to have commit rights there, so it would basically mean that
devs would not have commit rights to pymc-examples and/or resources repo.
**However**, this would become much more relevant and sensible if we had more
teams! (see next question).

> 2. do we want to create/formalize other teams?

Reviewer 2: There are not than many GitHub permission levels and not so many repositories for which we'd need teams anyway. It would quickly become redundant.

Author: New teams might not have anything to do with github. I personally would wait for teams to start organically, but I think it's good for people to know it's possible to do. See the details dropdown in the pr description for an example of possible evolution in a few years

- Jon Sedar
- Thomas Wiecki

## BDFL

The Project will have a BDFL (Benevolent Dictator for Life), who is currently
Chris Fonnesbeck. As Dictator, the BDFL has the authority to make all final
decisions for The Project. As Benevolent, the BDFL, in practice chooses to
defer that authority to the consensus of the community discussion channels and
the Steering Council. It is expected, and in the past has been the
case, that the BDFL will only rarely assert their final authority. Because
rarely used, we refer to BDFL’s final authority as a “special” or “overriding”
vote. When it does occur, the BDFL override typically happens in situations
where there is a deadlock in the Steering Council or if the Steering Council
asks the BDFL to make a decision on a specific matter. To ensure the
benevolence of the BDFL, The Project encourages others to fork the project if
they disagree with the overall direction the BDFL is taking. The BDFL is chair
of the Steering Council (see below) and may delegate their authority on a
particular decision or set of decisions to any other Council member at their

> discretion.

Reviewer 4: Long live the dictator!

they feel has been stale for too long and for which
informal consensus appears unlikely.
For a vote to be called, the target issue must be at least 2 months old.

To do so, they have to open a proposal issue ticket labeled "Council Vote".
The proposal issue should contain a link to the target issue and
a proposal on how to resolve it.
Proposals should include a statement making clear what it means to
"agree" or to "disagree".

Before voting starts, at least 3 days will be left for Core Contributors
to raise doubts about the proposal's phrasing, no extra discussion will
take place in the proposal issue.
Proposal issues should be locked from creation to prevent attracting
discussion from people not familiar with the decision process.

A vote is also called automatically whenever someone is nominated to
be added to the team.

The Steering Council can also call a vote on their own in order
to eject a Core contributor.

Upon ejecting a core contributor the council must publish an issue ticket,
or public document detailing the
* Violations
* Evidence if available
* Remediation plan (if necessary)
* Signatures majority of council members to validate correctness and accuracy

#### Voting

> * Each Council Member will vote either "Yes", "No", or "Neutral".

Reviewer 4: What's the purpose of a Neutral vote? Would you not simply abstain if you did not want to vote either way?

Author: Mostly to ensure everyone "votes" and we ensure nobody missed the notification/email or whatever and abstained "involuntarily". To a lesser extent because I expect voting to be async so with a neutral vote, even if we set a say 2-4 days time to vote it can finish early if everyone has voted.

* It is recommended that all Council Members expose their reasons when voting.
  "No" votes, however, must list the reasons for disagreement.
  Any "No" vote with no reason listed will be considered a "Neutral" vote.

> * An absence of vote is considered as "Neutral".

Reviewer 4: OK, this answers my question above. Might just want to call this "Abstain".

* Voting will remain open for at least 3 days.
* For the proposal to pass, at least 60% of the council must vote "Yes", and no more than 20% can vote "No".

For decisions about the project the Council will perform it directly on the proposal issue. For decisions about people, such as electing or ejecting Core Contributors, the Council will vote privately. However the decision will be posted publicly in an issue ticket.

> ### Vote of no conficence

Reviewer 7: ```suggestion
### Vote of no confidence
```

In exceptional circumstances, Council Members as well as Core Contributors
may remove a sitting council member via a vote of no confidence.
Core contributors can also call for a vote to remove the entire council
-- in which case, Council Members do not vote.
A no-confidence vote is triggered when a Core Contributor calls for one
publicly on an appropriate project communication channel,
and two other core team members second the proposal.
The initial call for a no-confidence vote must specify which type is intended
-- whether it is targeting a single member or the council as a whole.

The vote lasts for two weeks, and the people taking part in it vary:
* If this is a single-member vote called by Core contributors,
  both Council members and Core contributors vote,
  and the vote is deemed successful if at least two thirds of voters
  express a lack of confidence.
* If this is a whole-council vote, then it was necessarily called by
  Core contributors (since Council members can’t remove the whole Council)
  and only Core contributors vote.
  The vote is deemed successful if at least two thirds of voters

>   express a lack of confidence.

Reviewer 4: Does no-confidence as a council member also remove that member as a core contributor?

After voting:
* If a single-member vote on a council member succeeds, then that member is
  removed from the council and the resulting vacancy can be handled in the usual way.
* If a single-member vote on a core contributor succeeds, their permissions are
  revoked and would have to wait six months to be eligible for core contributor
  nomination again.
* If a whole-council vote succeeds, the council is dissolved and a new council election is triggered immediately.

### Election process
> Used when choosing the steering council and it's subcommittees

#### Nominations
* Nominations are taken over a single GitHub issue ticket over the course of 2 weeks.

> * Only Core Contributors may nominate folks

Reviewer 4: "... may nominate Council Members"

* Self Nominations are allowed
* At the conclusion of the 2 weeks, the list of nominations is posted on the ticket and this ticket is closed.

#### Voting

* Voting occurs over a period of at least 1 week, at the conclusion of the nominations.
  Voting is blind and mediated by either an application or a third party like NumFOCUS.
  Each voter can vote zero or more times, once per each candidate.
  As this is not about ranking but about capabilities,

>   voters vote on a yes/neutral/no basis per candidate -- “would I trust this person to lead ArviZ?”.

Reviewer 8: ```suggestion
  voters vote on a yes/neutral/no basis per candidate -- “would I trust this person to lead PyMC?”.
```

* Candidates are evaluated independently,
  each candidate having 60% or more of yes votes and less or

>   equal than 20% of no votes is chosen.

Reviewer 4: Seems like it would be easy to deadlock under this rule. So if 50% of members do not vote, you cannot be elected, even if there are no explicit "No" votes? Or am I misreading that?

Author: Then we would go directly to step 2. But I would consider 50% of core contributors not voting a huge red flag. I expect 100% of core contributors to vote. Recurrent contributors don't vote.

    requirements from being met.
* In the event of a tie there will be a runoff election for the tied candidates.
  To avoid further ties and discriminate more among the tied candidates,
  this vote will be held by Majority Judgment (MJ):
  for each candidate, voters judge their suitability for office as either
  "Excellent", "Very Good", "Good", "Acceptable", "Poor", or "Reject".
  Multiple candidates may be given the same grade by a voter.
  The candidate with the highest median grade is the winner.
* If more than one candidate has the same highest median-grade,
  the MJ winner is discovered by removing (one-by-one) any grades equal
  in value to the shared median grade from each tied candidate's total.
  This is repeated until only one of the previously tied candidates
  is currently found to have the highest median-grade.
* If ties are still present after this second round, the winner will be chosen at random. First we make a alphanumerically sorted list of the names in the tie. Then we will draw one prior predictive sample from a `pm.Categorical` distribution over the elements in the list to determine the winner.

> * At the conclusion of voting, all the results will be posted. And at least 24 hours will be left to challenge the election result in case there were suspicions of irregularities or the process had not been correctly carried out.

Reviewer 4: Who oversees the election? Should the council include an election officer?

Author: Someone external, it is mentioned above. In ArviZ last election I believe it was a numfocus representative cc @canyon289. The detailed log is here: https://github.com/arviz-devs/arviz/blob/main/elections/ArviZ_2020.md

Reviewer 7: NumFOCUS oversaw the election for us, and they were both great and kind about it

> * If ties are still present after this second round, the winner will be chosen at random. Each person tied will pick an integer number in the [1, 100] interval and send it privately to the third party mediating the election. After receiving all the numbers, said third party will draw a random integer from random.org. The person with the closest circular distance, defined as min(|a-b|, 100-|a-b|), will be selected. This process will be repeated as many times as necessary as there may be ties resulting from candidates choosing the same number.

Reviewer 2: This sounds a bit complicated. Why not sample the prior from a categorical distribution and pick the one with most samples?

Author: Sounds good, I copied that from ArviZ but I am not tied to that, I also hope we don't really get there but better safe than sorry. Can you provide a suggested wording?

Reviewer 2: ```suggestion
* If ties are still present after this second round, the winner will be chosen at random. First we make a alphanumerically sorted list of the names in the tie. Then we will draw one prior predictive sample from a `pm.Categorical` distribution over the elements in the list to determine the winner.
```

## Leaving the project
Any contributor can also voluntarily leave the project by notifying the community through a public means or by notifying the entire council. When doing so, they can add themselves to the alumni section below if desired.

> People who leave the project voluntarily can rejoin at any time

Reviewer 8: ```suggestion
People who leave the project voluntarily can rejoin at any time.
```

Decisions should be reached by consensus within the team with the Steering
Council and the BDFL acting if necessary.

The dedicated teams are work units with two main objectives: better
distributing the work related to The Project, and to better showcase all the task
involved in The Project to attract more diverse Contributors.

The PyMC project currently counts with the Developer and Documentation teams.
Team members can be part of one, some or none of these dedicated teams.

Team members are expected to participate and join these dedicated teams
organically, that is, the Steering Council will take part actively
in team assignments if they are part of a role change with the respective
nomination.

### Developer Team
The focus of the developer team is the probabilistic programming library
and flagship of The Project, [PyMC](https://github.com/pymc-devs/pymc).

For current members of the developer team, refer to the recurrent and
core contributor membership sections.

### Documentation Team
The focus of the documentation team is ensuring the PyMC library
is well documented, building and maintaining the infrastructure needed

> for that aim and making sure there are resources to learn

Reviewer 7: ```suggestion
for that aim, and making sure there are resources to learn
```

Bayesian statistics with PyMC.
It is not the goal nor responsibility of the Documentation team to
write all the documentation for the PyMC library.

For current members of the documentation team, refer to the recurrent and
core contributor membership sections.

### "No-team" tasks
All tasks related to the project that are not specifically listed in the
description of a dedicated team are the responsibility of the PyMC team
as a whole. At the time of writing, this includes but is not limited to:
fundraising, issue triaging, running PyMC related events like PyMCon or
sprints, outreach or presence on social networks.

### Team structure in practice
Our two teams are currently structured about GitHub centric tasks, so the
permissions on GitHub repositories is mapped to team membership and role
within the team. The team defines to which repositories the permissions
are given, the role defines the type of permissions given:

Role:
- Recurring Contributors are given triage permissions
- Core Contributors are given write permissions

Team:

> * Development team members are given permissions to [pymc](https://github.com/pymc-devs/pymc) repository

Reviewer 8: ```suggestion
* Development team members are given permissions to the [pymc](https://github.com/pymc-devs/pymc) repository.
```

> **Tier 1** = an institution with at least one Institutional Council Member

Reviewer 5: does this (potentially) make every steering council member's employer a tier 1 institution? (has that always been the case and I'm only noticing now?)

Author: from what I understand, yes, any employer of a steering council member is a potential tier 1 institutianal partner provided the duties of the council member include contributing to pymc and the steering council approves the partnership. And yes, that has always been the case, I have basically left everything related to institutional partners untouched.

> - Acknowledged on the PyMC websites, in talks and T-shirts.

Reviewer 7: Does this mean in every pymc related talk we are supposed to include Tier 1 partners?

Author: I think in every pymc related talk is impossible, and it can at most be pymc organized talk. But all these advantages probably need reviewing, I left this section mostly unmodified as I don't know enough about it. Would it be ok to sort this out in a follow-up PR?

Reviewer 7: Yea totally fine, this is a much less critical part versus the rest. Dont let it block a merge

- Ability to acknowledge their own funding sources on the PyMC
  websites, in talks and T-shirts.
- Unlimited participation in the annual Institutional Partners Workshop, held
  during the (planned) annual PyMC Project Retreat. This allows the
  Institutional Partner to invite as many of their own employees and funding
  sources and collaborators as they want, even if they are not project
  Contributors or Council Members.
- Ability to influence the project through the participation of their Council
  Member.

> - Council Members are invited to the bi-annual PyMC Developer Meeting.

Reviewer 7: Do we have a bi annual developer meeting?

Reviewer 4: That was the plan until COVID hit.

Reviewer 7: Hope we keep it the plan :) Covid ruined this last one for me but fingers crossed for the next one!

**Tier 2** = an institution with at least one Institutional Contributor

- Same benefits as Tier 1 level Partners, but:
- Only Institutional Contributors are invited to the Institutional Partners
  Workshop and bi-annual PyMC Developer Meeting

The PyMC project currently recognizes PyMC Labs as a Tier 1 Institutional Partner,

> with Thomas Wiecki and Adrian Seyboldt as their institutional contributors

Reviewer 5: this is wrong as written (where there is no steering council), but I think right as I am suggesting

Author: agreed, but we can fix this after the election I think, otherwise chances are we might have to do it anyways even if fixed now.

and council members.

## Sponsors

> Sponsors are Organizations that provide significant funding to the PyMC project

Reviewer 5: I don't have a good suggestion, but this section reads a little hostile towards sponsors? Maybe something like "Interested sponsors are encouraged to reach out to the Steering Council to arrange the sponsorship and recognition."

> either directly or by sponsoring PyMCon.

Reviewer 7: ```suggestion
directly.
```

Suggested proposal: By contributing to PyMCon you effectively are contriburing to PyMC project anyway so this is redundant
